packages/core/src/common/disposable.ts

```diff
@@ -47,6 +47,29 @@ Object.defineProperty(Disposable, 'NULL', {
     }
 });
 
+/**
+ * Utility for tracking a collection of Disposable objects, specifically for
+ * the sake of finding out when all items in the collection have been disposed.
+ *
+ * ```
+ * const dc = new DisposableCollection(myDisposables);
+ * dc.onDispose(() => {
+ *    console.log('All elements in the collection have been disposed');
+ * });
+ * ```
+ *
+ * You can add any number of Disposables to the collection at any time. You
+ * cannot query the contents of the collection. An element is automatically
+ * pruned from the collection when it is disposed.
+ *
+ * The collection will notify only once. It will continue to function
+ * in so far as accepting new Disposables and pruning them when they
+ * are disposed, but such activity will never result in another
+ * notification.
+ *
+ * Calling dispose() on the collection disposes all its elements. A notification
+ * is sent (if no notification was previously sent).
+ */
```

Comment:
Since we are already touching this, could I maybe ask you to add a test case to `disposable.spec.ts` that also covers the behavior in more detail?

```
import { expect, spy, use } from 'chai';
import { DisposableCollection, Disposable } from './disposable';
import * as spies from 'chai-spies';

use(spies);

describe('Disposables', () => {
    [...]
    it('onDispose is only called once on actual disposal of elements', () => {
        const onDispose = spy(() => { });
        const elementDispose = spy(() => { });

        const collection = new DisposableCollection();
        collection.onDispose(onDispose);

        // if the collection is empty 'onDispose' is not called
        collection.dispose();
        expect(onDispose).to.not.have.been.called();

        // 'onDispose' is called because we actually dispose an element
        collection.push(Disposable.create(elementDispose));
        collection.dispose();
        expect(elementDispose).to.have.been.called.once;
        expect(onDispose).to.have.been.called.once;

        // if the collection is empty 'onDispose' is not called and no further element is disposed
        collection.dispose();
        expect(elementDispose).to.have.been.called.once;
        expect(onDispose).to.have.been.called.once;

        // 'onDispose' is not called again even if we actually dispose an element
        collection.push(Disposable.create(elementDispose));
        collection.dispose();
        expect(elementDispose).to.have.been.called.twice;
        expect(onDispose).to.have.been.called.once;
    });
    [...]
```

Reply:
You bet

Reply:
... and you even added a test case of your own, thank you!